plutus-core/testlib/PlutusCore/Generators/PIR.hs

```diff
@@ -0,0 +1,1754 @@
+{-# LANGUAGE BangPatterns          #-}
+{-# LANGUAGE ConstraintKinds       #-}
+{-# LANGUAGE DataKinds             #-}
+{-# LANGUAGE DeriveAnyClass        #-}
+{-# LANGUAGE DeriveFunctor         #-}
+{-# LANGUAGE DerivingVia           #-}
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE ImportQualifiedPost   #-}
+{-# LANGUAGE LambdaCase            #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NumericUnderscores    #-}
+{-# LANGUAGE OverloadedStrings     #-}
+{-# LANGUAGE PartialTypeSignatures #-}
+{-# LANGUAGE PatternSynonyms       #-}
+{-# LANGUAGE RecordWildCards       #-}
+{-# LANGUAGE TemplateHaskell       #-}
+{-# LANGUAGE TupleSections         #-}
+{-# LANGUAGE TypeApplications      #-}
+{-# LANGUAGE TypeFamilies          #-}
+{-# LANGUAGE UndecidableInstances  #-}
+{-# LANGUAGE ViewPatterns          #-}
+{-# OPTIONS_GHC -Wno-name-shadowing #-}
+{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}
+{-# OPTIONS_GHC -Wno-partial-type-signatures #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+{-# OPTIONS_GHC -Wno-redundant-constraints #-}
+
+module PlutusCore.Generators.PIR where
+
+import Debug.Trace
+
+import Control.Applicative ((<|>))
+import Control.Arrow hiding ((<+>))
+import Control.Lens ((<&>))
+import Control.Monad.Except
+import Control.Monad.Reader
+import Control.Monad.Trans.Maybe
+
+import Data.Foldable
+import Data.List.NonEmpty (NonEmpty (..))
+import Data.Map (Map)
+import Data.Map qualified as Map
+import Data.Maybe
+import Data.Set (Set)
+import Data.Set qualified as Set
+import Data.String
+import GHC.Stack
+import Prettyprinter
+import Test.QuickCheck
+import Text.PrettyBy
+
+import PlutusCore.Default
+import PlutusCore.Name
+import PlutusCore.Normalize
+import PlutusCore.Quote (runQuoteT)
+import PlutusCore.Rename
+import PlutusIR
+import PlutusIR.Compiler
+import PlutusIR.Core.Instance.Pretty.Readable
+import PlutusIR.Error
+import PlutusIR.TypeCheck
+
+{- Note [Debugging generators that don't generate well-typed/kinded terms/types]
+    This module implements generators for well-typed terms and well-kinded types.
+    If you came here after a property like `prop_genTypeCorrect` failed and you
+    didn't get a minimal counterexample (because shrinking requries well-typed terms)
+    you need to use a different trick. One trick that often works is to add the well-typedness
+    check in the generators - e.g. in `genTerm` you can do something like this:
+    ```
+    genTerm ... = myCheck $ do
+      ...
+      where
+        myCheck gen = do
+          (trm, type) <- gen
+          if notConformingToExpectedInvariant trm type then
+            error (show trm ++ " " ++ show type)
+          else
+            return (trm, type)
+    ```
+-}
+
+-- | Term generators carry around a context to know
+-- e.g. what types and terms are in scope.
+type GenTm = ReaderT GenEnv Gen
+
+data GenEnv = GenEnv
+  { geSize               :: Int
```

Comment:
This is because you don't want to mix recursion control and size - if you used QuickCheck's builtin size you would get e.g. big integers near the root of the term and smaller integers near the leaves - not ideal. I'll write a little comment documenting this.

Reply:
Yikes. Sounds like an indictment of the whole concept of sized generators tbh.